### PR TITLE
allocations: test for overflow of requested size

### PR DESCRIPTION
**Pull request by ethomson**
Introduce some helper macros to test integer overflow from arithmetic and set error message appropriately.

We have a couple of places where we compute the size of the buffer we would like to allocate, but do not check that our size computation did not overflow.  I am suggesting a few helper macros here to assist with sanity checking our arithmetic.  The `GIT_ALLOC_OVERFLOW` macros simply test whether an overflow would occur if you were to add / multiply the operands before passing them to `malloc`, while the `GITERR_CHECK_ALLOC_*` macros actually set an error code and return `-1`, so they can be used alongside `GITERR_CHECK_ALLOC`.

Note that they actually call `giterr_set_oom`, this seems like a reasonable result when requesting more memory than is representable.

I tried to not add this in places where we could logically determine that we could never possibly overflow.  (Eg, `malloc(GIT_OID_SZ+1)` or something).  Still, I tried to keep these changes minimally intrusive - it's only a single additional line of code in most places.

While I was doing this, I also cleaned up some occurrences where we were not checking the return code for `malloc`, mostly in blame.  I also cleaned up some occurrences of backwards arguments to `calloc`.  Though this is not an Actual Problem, it seemed like the right thing to do.

### CODE REVIEW

**Comment by arrbee** [comment]
I like how this will increase robustness. :sparkles: It's like this is a professional library or something. :grinning:

I wonder, if we're going to make such a broad change, would it be cleaner just to change the type signature of `git__realloc` from `git__realloc(ptr, new_size)` to be like `calloc` and read `git__realloc(ptr, new_count, new_size)` and just do the overflow check internally? I think the big :+1: is that you wouldn't need the separate macro call in many places and would just have to add the check inside `realloc`, and the big :-1: is that this isn't how `realloc` works in standard C.

Along those lines, in the past I've considered whether we should add a `git_buf_grow_by(git_buf *buf, size_t amount_to_extend)` API that assumes you are adding to the existing length. Conceptually this fits with most uses and it can wrap some of the overflow logic. This won't eliminate nearly as much usage of your fine new macros (because `size + 1` still needs to be checked before it is added to `buf->size`), but it feels like an intuitive addition to the API and does streamline the base case of buffer growth.

Looking at the actual changes, it appears the most common pattern of buffer growth is to add `a * x + b` to the buffer length (where `a` is often just 1), so we could even bake that in and have `git_buf_grow_by` take 2 size inputs instead of just 1. That makes for a very strange API (why 2 size parameters?), but pragmatically it is by far the most common usage and would streamline many of the uses.

**Comment by carlosmn** [comment]
Changing the signature of `git__realloc()` isn't great, but the OpenBSD project recently added `reallocarray()` their libc which is the realloc equivalent to calloc, which also checks for overflow. We could introduce `git__reallocarray()` with number + size arguments which also checks for overflow.

That would also allow us to perform the overflow check inside the memory-allocating function instead of having to add it to each allocation (and would lead to a more direct diff, I think). Eventually we could go as far as not allowing `git__realloc()` but have to always rely on `git__reallocarray()`, but that's more of a long-term thing.

I do like the `git_buf_grow_by()` idea, which would again let us perform overflow checks inside the allocating function rather than doing it around it in the call sites.

I've wanted to change some backwards calloc args for a while; this seems like a pretty good opportunity to do so :)

**Comment by ethomson** [comment]
Yep, I think that both a `git_buf_grow_by` and a `git__reallocarray` are very sensible.  Thanks for the feedback here - going through all the allocations sort of warped my brain and I think I lost sight of the goal at the end.

Anyway, I will add both of those.  Though I don't think that it will eliminate all allocation arithmetic checking into its own functions and these macros will remain useful (just lesser used - though hopefully _a lot_ lesser).

**Comment by ethomson** [comment]
I introduced `git__reallocarray`, which is quite nice.  (Also `git__mallocarray` because an unnecessary `calloc` makes me sad, and `reallocarray(NULL, ...` makes me moreso.)  Also `git_buf_grow_by`.

There's still several places where we have several operands and so we need to test each arithmetic step for overflow.  Meh.  But short of some trickery that is not in my bag of tricks, I think this is an improvement.

**Comment by carlosmn** [comment]
The reallocarray stuff looks pretty good. reallocarray(NULL, ...) is a bit ugly but it's the safe way, the wrapper looks good.

There's a few places where you've added a call to `giterr_set_oom()` after the allocator returns `NULL` but the allocator functions already do this. I think some are left over from the ifs which checked for overflow, but they now just have reallocarray/mallocarray which makes it unnecessary.

File: src/common.h [hunk 1]
```diff
@@ -174,6 +174,28 @@ GIT_INLINE(void) git__init_structure(void *structure, size_t len, unsigned int v
 	GITERR_CHECK_VERSION(&(VERSION), _tmpl.version, #TYPE);	\
 	memcpy((PTR), &_tmpl, sizeof(_tmpl)); } while (0)
 
+/** Check for integer overflow from addition or multiplication */
+#define GIT_ALLOC_OVERFLOW_ADD(one, two) \
+	((one) + (two) < (one))
```
Comment (csware) [inline src/common.h]: IIRC, this is undefined in C and might not work as intended. See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=30475 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=30475#c8

**Comment by ethomson** [comment]
As @csware and @vmg astutely pointed out, you can't simply do the arithmetic and then check for overflow, since that's not defined on unsigned types in C.  `(a + b < a)` is obviously always false and an optimizing compiler can strip that away.

Instead, our naive test is now for [`(SIZE_MAX - a < b)`](http://c-faq.com/misc/intovf.html), but we introduce a new function to do this test for us, `git__add_sizet_overflow`, which looks similar to the [compiler builtins provided by clang and gcc](https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html) (but with libgit2-style arguments, the `out` param leads).

Since these also actually *do* the addition, adapt the macros (which are still useful to set an error code) to do the same.

The last commit here (`d33f2ab`) tries to detect and enable the compiler intrinsic for gcc and clang and use the appropriate one.  I'm pushing it up here but cannot test it until I'm in front of a machine that has a reasonably new clang (so please do not merge this yet).

**Comment by ethomson** [comment]
In the commits concurrent with my last comment, I had originally swapped the return code of the `add_overflow` functions compared to the gcc/clang implementation.  I fixed that error and now I think this is ready for review.

(With that error, every allocation failed, which illustrated some problems when there's an allocation failure in `git_libgit2_init`.  I'll investigate and open a pull request around this issue.)

**Comment by carlosmn** [comment]
In  f743492 you say you want to make the overflow checks more like clar's. Do you mean clang's?

File: src/pack-objects.c [hunk 2]
```diff
@@ -817,7 +832,9 @@ static int try_delta(git_packbuilder *pb, struct unpacked *trg,
 		trg_object->delta_data = NULL;
 	}
 	if (delta_cacheable(pb, src_size, trg_size, delta_size)) {
-		pb->delta_cache_size += delta_size;
+		if (git__add_uint64_overflow(&pb->delta_cache_size, pb->delta_cache_size, delta_size))
+			return -1;
+
 		git_packbuilder__cache_unlock(pb);
```
Comment (carlosmn) [inline src/pack-objects.c]: If the value overflows, we'd end up leaving the cache locked, no? Should we unlock unconditionally and return -1 if it overflowed?

**Comment by ethomson** [comment]
Haha, it's true, I did mean clang and not clar.  Reworded that, and pushed a change to the pack-objects behavior that you noted.  Thanks for the careful set of eyes.
